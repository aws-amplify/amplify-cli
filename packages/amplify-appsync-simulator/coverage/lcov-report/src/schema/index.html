
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/schema</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/schema</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">72.22% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>39/54</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">46.66% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>7/15</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">71.42% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>10/14</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">73.46% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>36/49</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="72.22" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 72%"></div><div class="cover-empty" style="width: 28%"></div></div>
	</td>
	<td data-value="72.22" class="pct medium">72.22%</td>
	<td data-value="54" class="abs medium">39/54</td>
	<td data-value="46.66" class="pct low">46.66%</td>
	<td data-value="15" class="abs low">7/15</td>
	<td data-value="71.42" class="pct medium">71.42%</td>
	<td data-value="14" class="abs medium">10/14</td>
	<td data-value="73.46" class="pct medium">73.46%</td>
	<td data-value="49" class="abs medium">36/49</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-04-03T17:51:02.795Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    