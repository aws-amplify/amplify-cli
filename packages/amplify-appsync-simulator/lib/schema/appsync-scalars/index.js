"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapSchema = exports.scalars = void 0;
const url_1 = require("url");
const graphql_1 = require("graphql");
const libphonenumber_js_1 = require("libphonenumber-js");
const net = __importStar(require("net"));
const graphql_iso_date_1 = require("graphql-iso-date");
const EMAIL_ADDRESS_REGEX = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
const phoneValidator = (ast, options) => {
    const { country = 'US' } = options;
    const { kind, value } = ast;
    if (kind !== graphql_1.Kind.STRING) {
        throw new graphql_1.GraphQLError(`Query error: Can only parse strings got a: ${kind}`, [ast]);
    }
    const isValid = (0, libphonenumber_js_1.isValidNumber)(value, country);
    if (!isValid) {
        throw new graphql_1.GraphQLError('Query error: Not a valid phone number', [ast]);
    }
    return value;
};
class AWSPhone extends graphql_1.GraphQLScalarType {
    constructor(options = { name: null, description: null }) {
        const { name, description } = options;
        super({
            name,
            description,
            serialize: (value) => {
                const ast = {
                    kind: graphql_1.Kind.STRING,
                    value,
                };
                return phoneValidator(ast, options);
            },
            parseValue: (value) => {
                const ast = {
                    kind: graphql_1.Kind.STRING,
                    value,
                };
                return phoneValidator(ast, options);
            },
            parseLiteral: (ast) => phoneValidator(ast, options),
        });
    }
}
const AWSDate = new graphql_1.GraphQLScalarType({
    name: 'AWSDate',
    description: graphql_iso_date_1.GraphQLDate.description,
    serialize(value) {
        return graphql_iso_date_1.GraphQLDate.serialize(value);
    },
    parseValue(value) {
        return graphql_iso_date_1.GraphQLDate.parseValue(value) ? value : undefined;
    },
    parseLiteral(value) {
        return graphql_iso_date_1.GraphQLDate.parseLiteral(value) ? value.value : undefined;
    },
});
const AWSTime = new graphql_1.GraphQLScalarType({
    name: 'AWSTime',
    description: graphql_iso_date_1.GraphQLTime.description,
    serialize(value) {
        return graphql_iso_date_1.GraphQLTime.serialize(value);
    },
    parseValue(value) {
        return graphql_iso_date_1.GraphQLTime.parseValue(value) ? value : undefined;
    },
    parseLiteral(value) {
        return graphql_iso_date_1.GraphQLTime.parseLiteral(value) ? value.value : undefined;
    },
});
const AWSDateTime = new graphql_1.GraphQLScalarType({
    name: 'AWSDateTime',
    description: graphql_iso_date_1.GraphQLDateTime.description,
    serialize(value) {
        return graphql_iso_date_1.GraphQLDateTime.serialize(value);
    },
    parseValue(value) {
        return graphql_iso_date_1.GraphQLDateTime.parseValue(value) ? value : undefined;
    },
    parseLiteral(value) {
        return graphql_iso_date_1.GraphQLDateTime.parseLiteral(value) ? value.value : undefined;
    },
});
const AWSTimestamp = new graphql_1.GraphQLScalarType({
    name: 'AWSTimestamp',
    description: 'The AWSTimestamp scalar type represents the number of seconds that have elapsed \
since 1970-01-01T00:00Z. Timestamps are serialized and deserialized as numbers. Negative values \
are also accepted and these represent the number of seconds till 1970-01-01T00:00Z.',
    serialize(value) {
        return graphql_1.GraphQLInt.serialize(value);
    },
    parseValue(value) {
        return graphql_1.GraphQLInt.parseValue(value) ? value : undefined;
    },
    parseLiteral(value) {
        if (value.kind !== graphql_1.Kind.INT) {
            throw new graphql_1.GraphQLError(`Can only validate integers but received: ${value.kind}`);
        }
        return Number.parseInt(value.value, 10);
    },
});
const validateIPAddress = (value) => {
    if (typeof value !== 'string') {
        throw new TypeError(`Value is not string: ${value}`);
    }
    if (net.isIPv4(value) || net.isIPv6(value)) {
        return value;
    }
    throw new TypeError(`Value is not a valid IP address: ${value}`);
};
const AWSIPAddress = new graphql_1.GraphQLScalarType({
    name: 'AWSIPAddress',
    description: 'The AWSIPAddress scalar type represents a valid IPv4 or IPv6 address string.',
    serialize(value) {
        return validateIPAddress(value);
    },
    parseValue(value) {
        return validateIPAddress(value);
    },
    parseLiteral(ast) {
        if (ast.kind !== graphql_1.Kind.STRING) {
            throw new graphql_1.GraphQLError(`Can only validate strings as IPv4 or IPv6 addresses but got a: ${ast.kind}`);
        }
        return validateIPAddress(ast.value);
    },
});
const parseJson = (value) => {
    if (typeof value !== 'string') {
        throw new graphql_1.GraphQLError(`Unable to parse ${value} as valid JSON.`);
    }
    try {
        return JSON.parse(value);
    }
    catch (error) {
        throw new TypeError(`Unable to parse ${value} as valid JSON.`);
    }
};
const AWSJSON = new graphql_1.GraphQLScalarType({
    name: 'AWSJSON',
    description: 'The AWSJSON scalar type represents a valid json object serialized as a string.',
    serialize(value) {
        return JSON.stringify(value);
    },
    parseValue(value) {
        return parseJson(value);
    },
    parseLiteral(ast) {
        if (ast.kind !== graphql_1.Kind.STRING) {
            throw new graphql_1.GraphQLError(`Unable to parse ${ast.kind} as valid JSON.`);
        }
        return parseJson(ast.value);
    },
});
const validateEmail = (value) => {
    if (typeof value !== 'string') {
        throw new TypeError(`Value is not string: ${value}`);
    }
    if (!EMAIL_ADDRESS_REGEX.test(value)) {
        throw new TypeError(`Value is not a valid email address: ${value}`);
    }
    return value;
};
const AWSEmail = new graphql_1.GraphQLScalarType({
    name: 'AWSEmail',
    description: 'A field whose value conforms to the standard internet email address format as specified in RFC822: https://www.w3.org/Protocols/rfc822/.',
    serialize: validateEmail,
    parseValue: validateEmail,
    parseLiteral(ast) {
        if (ast.kind !== graphql_1.Kind.STRING) {
            throw new graphql_1.GraphQLError(`Can only validate strings as email addresses but got a: ${ast.kind}`);
        }
        return validateEmail(ast.value);
    },
});
const parseUrlValue = (value) => (value ? new url_1.URL(value.toString()) : value);
const AWSURL = new graphql_1.GraphQLScalarType({
    name: 'AWSURL',
    description: 'A field whose value conforms to the standard URL format as specified in RFC3986: https://www.ietf.org/rfc/rfc3986.txt.',
    serialize(value) {
        if (value) {
            return new url_1.URL(value.toString()).toString();
        }
        return value;
    },
    parseValue: parseUrlValue,
    parseLiteral(ast) {
        if (ast.kind !== graphql_1.Kind.STRING) {
            throw new graphql_1.GraphQLError(`Can only validate strings as URLs but got a: ${ast.kind}`);
        }
        return parseUrlValue(ast.value);
    },
});
exports.scalars = {
    AWSJSON,
    AWSDate,
    AWSTime,
    AWSDateTime,
    AWSPhone: new AWSPhone({ name: 'AWSPhone', description: 'AWSPhone' }),
    AWSEmail,
    AWSURL,
    AWSTimestamp,
    AWSIPAddress,
};
function wrapSchema(schemaString) {
    const scalarStrings = Object.keys(exports.scalars)
        .map((scalarKey) => `scalar ${scalarKey}\n`)
        .join('');
    return scalarStrings + schemaString;
}
exports.wrapSchema = wrapSchema;
//# sourceMappingURL=index.js.map