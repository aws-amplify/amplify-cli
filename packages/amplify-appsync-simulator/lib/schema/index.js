"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateResolvers = void 0;
const graphql_1 = require("graphql");
const schema_1 = require("@graphql-tools/schema");
const appsync_scalars_1 = require("./appsync-scalars");
const directives_1 = require("./directives");
function generateResolvers(schema, resolversConfig = [], simulatorContext) {
    const appSyncScalars = new graphql_1.Source(Object.keys(appsync_scalars_1.scalars)
        .map((scalar) => `scalar ${scalar}`)
        .join('\n'), 'AppSync-scalar.json');
    const documents = [schema, appSyncScalars, (0, directives_1.getDirectiveTypeDefs)()].map((s) => (0, graphql_1.parse)(s));
    const doc = (0, graphql_1.concatAST)([...documents]);
    const resolvers = resolversConfig.reduce((acc, resolverConfig) => {
        const typeObj = acc[resolverConfig.typeName] || {};
        const fieldName = resolverConfig.fieldName;
        const typeName = resolverConfig.typeName;
        typeObj[resolverConfig.fieldName] = {
            resolve: async (source, args, context, info) => {
                const resolver = simulatorContext.getResolver(resolverConfig.typeName, resolverConfig.fieldName);
                try {
                    if (typeName !== 'Subscription') {
                        const res = await resolver.resolve(source, args, context, info);
                        return res;
                    }
                    else if (!source) {
                        const res = await resolver.resolve(source, args, context, info);
                        return res;
                    }
                    return source;
                }
                catch (e) {
                    context.appsyncErrors.push(e);
                }
                return undefined;
            },
            ...(typeName === 'Subscription'
                ? {
                    subscribe: (source, args, context) => {
                        if (context.appsyncErrors.length) {
                            throw new Error('Subscription failed');
                        }
                        return simulatorContext.asyncIterator(fieldName);
                    },
                }
                : {}),
        };
        return {
            ...acc,
            [resolverConfig.typeName]: typeObj,
        };
    }, { Subscription: {} });
    const defaultSubscriptions = generateDefaultSubscriptions(doc, resolversConfig, simulatorContext);
    if (Object.keys(defaultSubscriptions).length || Object.keys(resolvers.Subscription).length) {
        resolvers.Subscription = {
            ...defaultSubscriptions,
            ...resolvers.Subscription,
        };
    }
    else {
        delete resolvers.Subscription;
    }
    return (0, directives_1.transformSchemaWithDirectives)((0, schema_1.makeExecutableSchema)({
        typeDefs: doc,
        resolvers: {
            ...resolvers,
            ...appsync_scalars_1.scalars,
        },
    }), simulatorContext);
}
exports.generateResolvers = generateResolvers;
function generateDefaultSubscriptions(doc, configuredResolvers, simulatorContext) {
    const configuredSubscriptions = configuredResolvers.filter((cfg) => cfg.fieldName === 'Subscription').map((cfg) => cfg.typeName);
    const schema = (0, graphql_1.buildASTSchema)(doc);
    const subscriptionType = schema.getSubscriptionType();
    if (subscriptionType) {
        const f = schema.getType(subscriptionType.name);
        if (f) {
            const fields = f.getFields();
            return Object.keys(fields)
                .filter((sub) => !configuredSubscriptions.includes(sub))
                .reduce((acc, sub) => {
                const resolver = {
                    resolve: (data) => data,
                    subscribe: () => simulatorContext.asyncIterator(sub),
                };
                return { ...acc, [sub]: resolver };
            }, {});
        }
    }
    return {};
}
//# sourceMappingURL=index.js.map