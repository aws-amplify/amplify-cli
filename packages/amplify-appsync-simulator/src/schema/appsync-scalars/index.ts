import { URL } from 'url';
import { GraphQLInt, GraphQLScalarType, GraphQLError, Kind, StringValueNode, ValueNode } from 'graphql';
import { isValidNumber } from 'libphonenumber-js';

import { GraphQLDate, GraphQLTime, GraphQLDateTime } from 'graphql-iso-date';

const EMAIL_ADDRESS_REGEX = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
const IPV4_REGEX = /^(?:(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(?:\/(?:[0-9]|[1-2][0-9]|3[0-2]))?)$/;
const IPV6_REGEX = /^(?:(?:(?:[0-9A-Fa-f]{1,4}:){6}(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|::(?:[0-9A-Fa-f]{1,4}:){5}(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:[0-9A-Fa-f]{1,4})?::(?:[0-9A-Fa-f]{1,4}:){4}(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:(?:[0-9A-Fa-f]{1,4}:){0,1}[0-9A-Fa-f]{1,4})?::(?:[0-9A-Fa-f]{1,4}:){3}(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:(?:[0-9A-Fa-f]{1,4}:){0,2}[0-9A-Fa-f]{1,4})?::(?:[0-9A-Fa-f]{1,4}:){2}(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:(?:[0-9A-Fa-f]{1,4}:){0,3}[0-9A-Fa-f]{1,4})?::[0-9A-Fa-f]{1,4}:(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:(?:[0-9A-Fa-f]{1,4}:){0,4}[0-9A-Fa-f]{1,4})?::(?:[0-9A-Fa-f]{1,4}:[0-9A-Fa-f]{1,4}|(?:(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}(?:0?0?[0-9]|0?[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5]))|(?:(?:[0-9A-Fa-f]{1,4}:){0,5}[0-9A-Fa-f]{1,4})?::[0-9A-Fa-f]{1,4}|(?:(?:[0-9A-Fa-f]{1,4}:){0,6}[0-9A-Fa-f]{1,4})?::)(?:\/(?:0?0?[0-9]|0?[1-9][0-9]|1[01][0-9]|12[0-8]))?)$/;

// Some of the custom scalars in this file are inspired by the graphql-scalars npm module.

const phoneValidator = (ast, options) => {
  const { country = 'US' } = options;
  const { kind, value } = ast;
  if (kind !== Kind.STRING) {
    throw new GraphQLError(`Query error: Can only parse strings got a: ${kind}`, [ast]);
  }

  let isValid = isValidNumber(value, country);
  if (!isValid) {
    throw new GraphQLError('Query error: Not a valid phone number', [ast]);
  }

  return value;
};

class AWSPhone extends GraphQLScalarType {
  constructor(options = { name: null, description: null }) {
    const { name, description } = options;
    super({
      name,
      description,
      serialize: value => {
        const ast = {
          kind: Kind.STRING,
          value,
        };
        return phoneValidator(ast, options);
      },
      parseValue: value => {
        const ast = {
          kind: Kind.STRING,
          value,
        };
        return phoneValidator(ast, options);
      },
      parseLiteral: ast => phoneValidator(ast, options),
    });
  }
}

const AWSDate = new GraphQLScalarType({
  name: 'AWSDate',
  description: GraphQLDate.description,
  serialize(value) {
    return GraphQLDate.serialize(value);
  },
  parseValue(value) {
    return GraphQLDate.parseValue(value) ? value : undefined;
  },
  parseLiteral(value) {
    return GraphQLDate.parseLiteral(value) ? (value as StringValueNode).value : undefined;
  },
});

const AWSTime = new GraphQLScalarType({
  name: 'AWSTime',
  description: GraphQLTime.description,
  serialize(value) {
    return GraphQLTime.serialize(value);
  },
  parseValue(value) {
    return GraphQLTime.parseValue(value) ? value : undefined;
  },
  parseLiteral(value) {
    return GraphQLTime.parseLiteral(value) ? (value as StringValueNode).value : undefined;
  },
});

const AWSDateTime = new GraphQLScalarType({
  name: 'AWSDateTime',
  description: GraphQLDateTime.description,
  serialize(value) {
    return GraphQLDateTime.serialize(value);
  },
  parseValue(value) {
    return GraphQLDateTime.parseValue(value) ? value : undefined;
  },
  parseLiteral(value) {
    return GraphQLDateTime.parseLiteral(value) ? (value as StringValueNode).value : undefined;
  },
});

const AWSTimestamp = new GraphQLScalarType({
  name: 'AWSTimestamp',
  description:
    'The AWSTimestamp scalar type represents the number of seconds that have elapsed \
since 1970-01-01T00:00Z. Timestamps are serialized and deserialized as numbers. Negative values \
are also accepted and these represent the number of seconds till 1970-01-01T00:00Z.',
  serialize(value) {
    return GraphQLInt.serialize(value);
  },
  parseValue(value) {
    return GraphQLInt.parseValue(value) ? value : undefined;
  },
  parseLiteral(value: ValueNode) {
    if (value.kind !== Kind.INT) {
      throw new GraphQLError(`Can only validate integers but received: ${value.kind}`);
    }

    return Number.parseInt(value.value, 10);
  },
});

// Unified the code for both types from graphql-scalars library.
const validateIPAddress = value => {
  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${value}`);
  }
  if (!IPV4_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid IPv4 address: ${value}`);
  }
  if (!IPV6_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid IPv6 address: ${value}`);
  }
  return value;
};

const AWSIPAddress = new GraphQLScalarType({
  name: 'AWSIPAddress',
  description: 'The AWSIPAddress scalar type represents a valid IPv4 or IPv6 address string.',
  serialize(value) {
    return validateIPAddress(value);
  },
  parseValue(value) {
    return validateIPAddress(value);
  },
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Can only validate strings as IPv4 or IPv6 addresses but got a: ${ast.kind}`);
    }
    return validateIPAddress(ast.value);
  },
});

const parseJson = (value: string) => {
  if (typeof value !== 'string') {
    throw new GraphQLError(`Unable to parse ${value} as valid JSON.`);
  }

  try {
    return JSON.parse(value);
  } catch (error) {
    throw new TypeError(`Unable to parse ${value} as valid JSON.`);
  }
};

const AWSJSON = new GraphQLScalarType({
  name: 'AWSJSON',
  description: 'The AWSJSON scalar type represents a valid json object serialized as a string.',
  serialize(value) {
    return JSON.stringify(value);
  },
  parseValue(value) {
    return parseJson(value);
  },
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Unable to parse ${ast.kind} as valid JSON.`);
    }

    return parseJson(ast.value);
  },
});

const validateEmail = value => {
  if (typeof value !== 'string') {
    throw new TypeError(`Value is not string: ${value}`);
  }

  if (!EMAIL_ADDRESS_REGEX.test(value)) {
    throw new TypeError(`Value is not a valid email address: ${value}`);
  }

  return value;
};

const AWSEmail = new GraphQLScalarType({
  name: 'AWSEmail',
  description:
    'A field whose value conforms to the standard internet email address format as specified in RFC822: https://www.w3.org/Protocols/rfc822/.',
  serialize: validateEmail,
  parseValue: validateEmail,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Can only validate strings as email addresses but got a: ${ast.kind}`);
    }

    return validateEmail(ast.value);
  },
});

const parseUrlValue = value => (value ? new URL(value.toString()) : value);

const AWSURL = new GraphQLScalarType({
  name: 'AWSURL',
  description: 'A field whose value conforms to the standard URL format as specified in RFC3986: https://www.ietf.org/rfc/rfc3986.txt.',
  serialize(value) {
    if (value) {
      return new URL(value.toString()).toString();
    }

    return value;
  },
  parseValue: parseUrlValue,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(`Can only validate strings as URLs but got a: ${ast.kind}`);
    }

    return parseUrlValue(ast.value);
  },
});

export const scalars = {
  AWSJSON,
  AWSDate,
  AWSTime,
  AWSDateTime,
  AWSPhone: new AWSPhone({ name: 'AWSPhone', description: 'AWSPhone' }),
  AWSEmail,
  AWSURL,
  AWSTimestamp,
  AWSIPAddress,
};

export function wrapSchema(schemaString) {
  const scalarStrings = Object.keys(scalars)
    .map(scalarKey => `scalar ${scalarKey}\n`)
    .join('');

  return scalarStrings + schemaString;
}
