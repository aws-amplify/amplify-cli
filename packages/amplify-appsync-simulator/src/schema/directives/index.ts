export { AwsSubscribe } from './aws-subscribe';
