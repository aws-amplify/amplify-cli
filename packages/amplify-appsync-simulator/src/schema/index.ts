import { buildASTSchema, concatAST, DocumentNode, GraphQLObjectType, parse, Source } from 'graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { AmplifyAppSyncSimulator } from '..';
import { AppSyncSimulatorPipelineResolverConfig, AppSyncSimulatorUnitResolverConfig } from '../type-definition';
import { scalars } from './appsync-scalars';
import { getDirectiveTypeDefs, transformSchemaWithDirectives } from './directives';

export function generateResolvers(
  schema: Source,
  resolversConfig: (AppSyncSimulatorUnitResolverConfig | AppSyncSimulatorPipelineResolverConfig)[] = [],
  simulatorContext: AmplifyAppSyncSimulator,
) {
  const appSyncScalars = new Source(
    Object.keys(scalars)
      .map((scalar) => `scalar ${scalar}`)
      .join('\n'),
    'AppSync-scalar.json',
  );

  const documents = [schema, appSyncScalars, getDirectiveTypeDefs()].map((s) => parse(s));
  const doc = concatAST([...documents]);

  const resolvers = resolversConfig.reduce(
    (acc, resolverConfig) => {
      const typeObj = acc[resolverConfig.typeName] || {};
      const fieldName = resolverConfig.fieldName;
      const typeName = resolverConfig.typeName;
      typeObj[resolverConfig.fieldName] = {
        resolve: async (source, args, context, info) => {
          const resolver = simulatorContext.getResolver(resolverConfig.typeName, resolverConfig.fieldName);
          try {
            // Mutation and Query
            if (typeName !== 'Subscription') {
              const res = await resolver.resolve(source, args, context, info);
              return res;
            } else if (!source) {
              // Subscription at connect time
              const res = await resolver.resolve(source, args, context, info);
              return res;
            }
            // subscription at publish time. No filtering
            return source;
          } catch (e) {
            context.appsyncErrors.push(e);
          }
          return undefined;
        },
        ...(typeName === 'Subscription'
          ? {
              subscribe: (source, args, context) => {
                // Connect time error. Not allowing subscription
                if (context.appsyncErrors.length) {
                  throw new Error('Subscription failed');
                }
                return simulatorContext.asyncIterator(fieldName);
              },
            }
          : {}),
      };
      return {
        ...acc,
        [resolverConfig.typeName]: typeObj,
      };
    },
    { Subscription: {} },
  );
  const defaultSubscriptions = generateDefaultSubscriptions(doc, resolversConfig, simulatorContext);

  if (Object.keys(defaultSubscriptions).length || Object.keys(resolvers.Subscription).length) {
    resolvers.Subscription = {
      ...defaultSubscriptions,
      ...resolvers.Subscription,
    };
  } else {
    // When there are no subscriptions in the doc, don't include subscription resolvers
    delete resolvers.Subscription;
  }

  return transformSchemaWithDirectives(
    makeExecutableSchema({
      typeDefs: doc,
      resolvers: {
        ...resolvers,
        ...scalars,
      },
    }),
    simulatorContext,
  );
}

function generateDefaultSubscriptions(
  doc: DocumentNode,
  configuredResolvers: (AppSyncSimulatorUnitResolverConfig | AppSyncSimulatorPipelineResolverConfig)[],
  simulatorContext: AmplifyAppSyncSimulator,
) {
  const configuredSubscriptions = configuredResolvers.filter((cfg) => cfg.fieldName === 'Subscription').map((cfg) => cfg.typeName);
  const schema = buildASTSchema(doc);
  const subscriptionType = schema.getSubscriptionType();
  if (subscriptionType) {
    const f = schema.getType(subscriptionType.name) as GraphQLObjectType;
    if (f) {
      const fields = f.getFields();
      return Object.keys(fields)
        .filter((sub) => !configuredSubscriptions.includes(sub))
        .reduce((acc, sub) => {
          const resolver = {
            resolve: (data) => data,
            subscribe: () => simulatorContext.asyncIterator(sub),
          };
          return { ...acc, [sub]: resolver };
        }, {});
    }
  }
  return {};
}
