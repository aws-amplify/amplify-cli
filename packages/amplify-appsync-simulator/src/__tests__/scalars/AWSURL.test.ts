import { URL } from 'url';
import { scalars } from '../../schema/appsync-scalars';

describe('AWSURL parse', () => {
  it('Returns falsy values unchanged', () => {
    expect(scalars.AWSURL.parseValue(0)).toEqual(0);
  });

  it('Returns valid URL objects', () => {
    const parsed = new URL('http://www.amazon.com');
    expect(scalars.AWSURL.parseValue('http://www.amazon.com')).toEqual(parsed);
  });

  it('Should reject an invalid URL', () => {
    function serialize() {
      scalars.AWSURL.parseValue('invalid-url');
    }
    expect(serialize).toThrowErrorMatchingSnapshot();
  });
});

describe('AWSURL serialize', () => {
  it('Returns falsy values unchanged', () => {
    expect(scalars.AWSURL.serialize(0)).toEqual(0);
  });

  it('Returns valid URLs', () => {
    expect(scalars.AWSURL.serialize('http://www.amazon.com')).toEqual('http://www.amazon.com/');
  });

  it('Should reject an invalid URL', () => {
    function serialize() {
      scalars.AWSURL.serialize('invalid-url');
    }
    expect(serialize).toThrowErrorMatchingSnapshot();
  });
});
