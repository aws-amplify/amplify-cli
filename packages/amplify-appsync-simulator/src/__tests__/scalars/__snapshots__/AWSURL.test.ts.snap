// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWSURL parse Should reject an invalid URL 1`] = `"Invalid URL: invalid-url"`;

exports[`AWSURL serialize Should reject an invalid URL 1`] = `"Invalid URL: invalid-url"`;
